### ISSUE TITLE
Tooltip text doesn't change when tooltipContent changes

### ISSUE THREAD

**Issue by cdl** [sir-dunxalot/ember-tooltips#5]
I have the following element inside a component:

```hbs
<a class="btn btn--grade has-tooltip"
    {{action "grade"}}
    data-tooltip-content={{gradeTooltipText}}>
  {{!-- button content... --}}
</a>
```

The component, on `didInsertElement`, calls `this.renderChildTooltips();`, which works great. However, **when the `data-tooltip-content` attribute changes, the content of the tooltip itself doesn't change**.

Subsequent calls to `this.renderChildTooltips()` doesn't re-render the existing tooltip - it just creates another tooltip with the new content, the old one still existing.

Is there a method I can call to destroy and re-render (or update) all tooltips? Or is this just a case that isn't covered yet?

**Comment by sir-dunxalot** [comment]
I think this is very valid, especially with the nature of promises.

The [tooltip API doc are available here](https://github.com/darsain/tooltip/wiki/Tooltip). There is a `Tooltip.detach()` method, which might suit our needs here. However, better still is the `Tooltip.content()` method, which changes the tooltip content - so it's definitely possible.

One blocker here is that whilst the tooltip is referenced as `tooltip` on components and views, the tooltips rendered using `renderChildTooltips` are not saved for later reference. This shouldn't be too hard to do. This would probably require each tooltip to have a unique ID so you can change to correct one.

Perhaps `renderChildTooltips()` and associated functionality could even be moved into its own mixin.

I'd be more than happy to accept a PR for this as it will probably take me a couple of weeks to get around to it.

**Comment by jamesarosen** [comment]
`renderTooltip` already [sets](https://github.com/sir-dunxalot/ember-tooltips/blob/3faafa5105d3f46cc58480aadfdc845081f0d772/addon/mixins/components/tooltips.js#L149) `tooltip` to the `Tooltip` instance.

Thus, the simplest thing that could _possibly_ work would be
```js
_updateTooltipContent: Ember.observes('tooltipContent', function() {
  const tooltip = this.get('tooltip');
  if (tooltip == null) { return; }

  tooltip.content(this.get('tooltipContent'));
})
```

I don't like that because it uses an Observer. It also might not work if the component that has `tooltipContent` differs from the one that has `tooltip`, which might happen with `{{#tooltip-on-parent}}`; I'm not sure

A more "correct" approach would be
```js
didReceiveAttrs({ newAttrs }) {
  const tooltip = this.get('tooltip');
  if (tooltip == null) { return; }

  tooltip.content(Ember.get(newAttrs, 'tooltipContent'));
  return this._super.apply(this, arguments);
})
```

But I don't like that because if a child class overrides `didReceiveAttrs` and forgets to call `_super`, the behavior gets erased.

**Comment by jamesarosen** [comment]
My current solution:
```js
_rerenderTooltip: on('didInsertElement', 'didUpdate', function() {
  const content = this.get('tooltipContent') || '';

  if (content.trim() === '') { return this.destroyTooltip(); }

  let tooltip = this.get('tooltip');
  if (!tooltip) { tooltip = this.renderTooltip(); }
  tooltip.content(content);
})
```

**Comment by sir-dunxalot** [comment]
Sorry for not realizing this sooner but #20 was recently merged into the Ember 2.0-compatible branch. Would this solve the issues you're having? Also, I'm open to another PR that would only add the observer in #20 when the Ember version is less than one that would support your above solution, which is more forward-compatible.

**Comment by jamesarosen** [comment]
#20 will only work if `tooltipContent` is truthy when `didInsertElement` fires. If it starts out `null` and then changes to `true`, `this.get('tooltip')` will be `undefined` when the observer fires.

**Comment by sir-dunxalot** [comment]
Ok, I see. Version 0.5.0 and above only supports Ember 2.0 and above so you'd be fine to use the `didUpdate` event without needing to support older Ember versions. Of course, if you'd like to make it backwards compatible that's not a bad thing.

Let me know if you need any other info or pointers.

**Comment by SirZach** [comment]
@sir-dunxalot I need some advice on updating dynamic content within the `tooltip-on-parent` component. 

```hbs
{{#tooltip-on-parent event='click'}}
  {{!-- promise represents a PromiseObectProxy --}}
  {{#if promise.isPending}}
    {{spinner}}
  {{/if}}
  {{#if promise.isFulfilled}}
    {{!-- show all the content surrounding the promise --}}
  {{/if}}
{{/tooltip-on-parent}}
```

Do I need to add an observer on the `tooltip-on-parent` component?

**Comment by sir-dunxalot** [comment]
Hi Zach, sorry for the slow response. The `tooltip-on-parent` is, unfortunately not set up well for rerendering.

In order to change the tooltip's content in the parent view you would need to get the tooltip object and call the method like follows:

```js
this.get('tooltip').content('The new tooltip text');
```

You could do this inside the observer you mentioned, or on `didUpdateAttrs` if you're using a newer version of Ember.

**Comment by sir-dunxalot** [comment]
And to answer your other question, the observer would be placed on the parent component. In your case, whatever component the `{{tooltip-on-parent}}` is in the layout of.

**Comment by Kinjeiro** [comment]
Hi! 
@SirZach @sir-dunxalot, i fix the problem in pull-request
https://github.com/sir-dunxalot/ember-tooltips/pull/66

**Comment by sir-dunxalot** [comment]
This has been fixed with release `1.0.0`: https://github.com/sir-dunxalot/ember-tooltips/releases/tag/1.0.0

@Kinjeiro We'll still try and merge your PR in to make a 0.7.1 patch.
